prow/cmd/deck/abort.go

```diff
@@ -0,0 +1,104 @@
+	ktypes "k8s.io/apimachinery/pkg/types"
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	prowv1 "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
+	"k8s.io/test-infra/prow/githuboauth"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+func handleAbort(prowJobClient prowv1.ProwJobInterface, cfg authCfgGetter, goa *githuboauth.Agent, ghc githuboauth.AuthenticatedUserIdentifier, cli deckGitHubClient, pluginAgent *plugins.ConfigAgent, log *logrus.Entry) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.TODO()
+		name := r.URL.Query().Get("prowjob")
+		l := log.WithField("prowjob", name)
+		if name == "" {
+			http.Error(w, "request did not provide the 'prowjob' query parameter", http.StatusBadRequest)
+			return
+		}
+		pj, err := prowJobClient.Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			http.Error(w, fmt.Sprintf("ProwJob not found: %v", err), http.StatusNotFound)
+			if !kerrors.IsNotFound(err) {
+				// admins only care about errors other than not found
+				l.WithError(err).Warning("ProwJob not found.")
+			}
+			return
+		}
+		switch r.Method {
+		case http.MethodPost:
+			if pj.Status.State != prowapi.TriggeredState && pj.Status.State != prowapi.PendingState {
+				http.Error(w, fmt.Sprintf("Cannot abort job with state: %q", pj.Status.State), http.StatusBadRequest)
+				l.Debug("Cannot abort job with state")
+				return
+			}
+			// Using same permission validation as rerun, could be future work to add validation
+			// unique to Abort
+			allowed, user, err, code := isAllowedToRerun(r, cfg, goa, ghc, *pj, cli, pluginAgent, l)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("Could not verify if allowed to abort: %v", err), code)
+				l.WithError(err).Debug("Could not verify if allowed to abort")
+			}
+			l = l.WithField("allowed", allowed)
+			l.Info("Attempted abort")
+			if !allowed {
+				http.Error(w, "You don't have permission to abort this job", http.StatusUnauthorized)
+				l.Debug("Error writing to abort response.")
+				return
+			}
+			var abortDescription string
+			if len(user) > 0 {
+				abortDescription = fmt.Sprintf("%v successfully aborted %v.", user, name)
+			} else {
+				abortDescription = fmt.Sprintf("Successfully aborted %v.", name)
+			}
+			pj.Status.State = prowapi.AbortedState
+			pj.Status.Description = abortDescription
+			jsonPJ, err := json.Marshal(pj)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("Error marshal source job: %v", err), http.StatusInternalServerError)
+				l.WithError(err).Errorf("Error marshal source job")
+			}
+			pj, err := prowJobClient.Patch(ctx, pj.Name, ktypes.MergePatchType, jsonPJ, metav1.PatchOptions{})
+			if err != nil {
+				http.Error(w, fmt.Sprintf("Could not patch aborted job: %v", err), http.StatusInternalServerError)
+				l.WithError(err).Errorf("Could not patch aborted job")
+			}
+			l.Info(abortDescription)
+			// Marshalls pj and adds it to body of response
```

Comment:
does the consumer of the response expect anything from the prowjob struct other than the status code? if not I don't think we need to include the entire prowjob struct as body